### PR TITLE
Fix email parsing

### LINKED ISSUE #202: [PE-D][Tester C] Email validity checking inconsistent with error message for local-part
The error message when keying in an invalid email states 
```
The local-part should only contain alphanumeric characters and these special characters, excluding the parentheses, (+_.-). The local-part may not start or end with any special characters.
```
However, when adding a new person or editing the email field of a person, when you input two special characters in a row in the `local-part`, the email is not accepted, even though it satisfies the requirements given in the error message. This might be difficult for a user if they wish to add such an email. One improvement would be to change the error message or the validity checking, so that they are consistent.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: rocketninja7/ped#3

### LINKED ISSUE #204: [PE-D][Tester C] Email validity checking inconsistent with error message for domain label
The error message when keying in an invalid email states 
```
The domain name is made up of domain labels separated by periods.
The domain name must:
    - end with a domain label at least 2 characters long
    - have each domain label start and end with alphanumeric characters
    - have each domain label consist of alphanumeric characters, separated only by hyphens, if any.
```
However, when adding a new person or editing the email field of a person, when you provide an email with a domain that ends with a domain label of one alphanumeric character, followed by a hyphen, then another alphanumeric character, it satisfies all the conditions listed above, yet it is rejected. This might be difficult for a user if someone's company has a domain that falls into this category. One improvement would be to change the error message or the validity checking, so that they are consistent.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: rocketninja7/ped#4

### PR DESCRIPTION
**Pull request by xavierpok**
Closes #202 
Closes #204 

### **Local part**
The regex is still mostly impenetrable without taking a moment to figure out what it actually means but should be a bit better

Regex in this PR for local part will become 
`^ALPHANUMERIC((SPECIAL_CHAR|ALPHANUMERIC)*ALPHANUMERIC)?`

Which roughly means that it must start with an alphanumeric (w/o underscore) character, then (optional) have zero or more of either (special char or alphanumeric) followed by one alphanumeric.

This should be right, though I invite anyone who's reviewing to double-check that this makes sense.

Test cases updated to reflect this new behaviour. There was a test case for this exact scenario that was supposed to fail, in contrast to advertised behaviour.

### **Domain part**
regex still similarly quite bad, but works properly now.
No real way to make it easy to read as far as I can see

Considered splitting up the suffix (`+` vs `*`) but that kind of ended up making it about as bad to read because the fact that you're modifying number of times you're repeating the last unit only obfuscated.

Also added comments roughly describing the logic of the regex portions.

Test cases also updated.

### CODE REVIEW

**Comment by xavierpok** [comment]
Converted to draft to address #204

File: src/main/java/connexion/model/person/Email.java [hunk 1]
```diff
@@ -9,7 +9,7 @@
  */
 public class Email implements PersonDetailField<String> {
 
-    private static final String SPECIAL_CHARACTERS = "+_.-";
+    private static final String SPECIAL_CHARACTERS = "[+_.-]+"; // one or more special characters as defined
     public static final String MESSAGE_CONSTRAINTS = "Emails should be of the format local-part@domain "
             + "and adhere to the following constraints:\n"
             + "1. The local-part should only contain alphanumeric characters and these special characters, excluding "
```
Comment (imkwokyong) [inline src/main/java/connexion/model/person/Email.java:15]: image
might need to fix the ([+_.-]+) part cause it's in regex form and might confuse the users with the square brackets and the "+" sign behind

**Review by imkwokyong — CHANGES_REQUESTED** [review]
A nit about the error message

File: src/main/java/connexion/model/person/Email.java (hunk 1 above)
Comment (xavierpok) [reply src/main/java/connexion/model/person/Email.java:18]: Done!

**Review by imkwokyong — APPROVED** [review]
LGTM!

**Pull request merged**